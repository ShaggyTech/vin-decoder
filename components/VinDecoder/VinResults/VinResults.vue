<script lang="ts">
/* Composition API */
import { defineComponent, ref, watch, PropType } from '@vue/composition-api';
/* Utility Methods */
import { filterResults } from '@/utils/filterResults';
/* Types */
import { ResultsObjectType } from '@/types';

export default defineComponent({
  name: 'VinResults',
  inheritAttrs: false,
  props: {
    rawResults: {
      required: false,
      type: Object as PropType<ResultsObjectType | null>,
      default: null
    },
    loading: {
      required: false,
      type: Boolean as PropType<boolean>,
      default: false
    },
    transition: {
      required: false,
      type: String as PropType<string>,
      default: 'slide-y-transition'
    },
    tableHeight: {
      required: false,
      type: [Number, String] as PropType<number | string>,
      default: 300
    },
    maxWidth: {
      required: false,
      type: [Number, String] as PropType<number | string>,
      default: 600
    }
  },

  setup(props) {
    const results = ref<ResultsObjectType | null>(null);

    watch(
      () => props.rawResults,
      (newRawResults: ResultsObjectType | null): void => {
        results.value = newRawResults ? filterResults(newRawResults) : null;
      }
    );

    return {
      results
    };
  }
});
</script>

<template>
  <component
    :is="transition !== 'None' ? `v-${transition}` : 'div'"
    hide-on-leave
  >
    <v-skeleton-loader
      v-if="loading"
      :max-width="maxWidth"
      type="card-heading, text@7"
    >
    </v-skeleton-loader>
    <v-card
      v-else-if="results && rawResults"
      v-bind="$attrs"
      :max-width="maxWidth"
      raised
    >
      <v-card-title class="mx-auto px-2 title">
        <v-divider></v-divider>
        <span class="mx-auto px-4">
          {{ results.ModelYear }} {{ results.Make }} {{ results.Model }}
          {{ results.Series }}
        </span>
        <v-divider></v-divider>
      </v-card-title>
      <v-card-subtitle class="mx-auto px-2 title text-center">
        {{ results.VIN }}
        <v-divider></v-divider>
      </v-card-subtitle>
      <v-card-text>
        <v-simple-table
          fixed-header
          :height="tableHeight"
          :max-width="maxWidth"
        >
          <template v-slot:default>
            <thead>
              <tr>
                <th class="text-left">Name</th>
                <th class="text-left">Value</th>
              </tr>
            </thead>
            <tbody>
              <tr v-for="(value, key, index) in results" :key="index">
                <td>{{ key }}</td>
                <td>{{ value }}</td>
              </tr>
            </tbody>
          </template>
        </v-simple-table>
      </v-card-text>
    </v-card>
  </component>
</template>
