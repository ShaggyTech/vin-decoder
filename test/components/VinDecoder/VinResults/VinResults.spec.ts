// import flushPromises from 'flush-promises';

import { shallowMount } from '@vue/test-utils';

import { mockRawResults } from '../../../__mocks__/mockDecodeVinValuesExtendedResults';
import VinResults from '@/components/VinDecoder/VinResults/VinResults.vue';

const factory = (options: object) => {
  return shallowMount(VinResults, {
    ...options
  });
};

describe('VinResults Component Tests', () => {
  test('component is hidden if rawResults prop is empty', () => {
    const wrapper = factory({
      sync: false
    });

    expect(wrapper.vm.$el.textContent).toEqual('');
    expect(wrapper.element).toMatchSnapshot();
  });

  test('if loading prop is true, a skeleton loader is shown in place of the results card', () => {
    const wrapper = factory({
      sync: false,
      propsData: {
        loading: true
      }
    });

    expect(wrapper.find('.results-loader').exists()).toBe(true);
    expect(wrapper.element).toMatchSnapshot();
  });

  test('given rawResults, component renders the correct title and subtitle', () => {
    const wrapper = factory({
      sync: false,
      propsData: {
        rawResults: mockRawResults
      }
    });

    const { Make, Model, ModelYear, VIN } = mockRawResults;
    const titleItems = { Make, Model, ModelYear };

    const titleText = wrapper.find('.results-card__title').text();
    Object.keys(titleItems).forEach(key =>
      expect(titleText).toContain(titleItems[key])
    );

    const subtitleText = wrapper.find('.results-card__subtitle').text();
    expect(subtitleText).toContain(`${VIN}`);

    expect(wrapper.element).toMatchSnapshot();
  });

  test('given rawResults, nullish or non-applicable values are filtered out of the results object ref', async () => {
    const wrapper = factory({
      sync: false,
      propsData: {
        rawResults: mockRawResults
      }
    });

    const { Make, Model, ModelYear, VIN } = mockRawResults;
    const validResults = { Make, Model, ModelYear, VIN };

    await wrapper.vm.$nextTick();

    const resultsRef = wrapper.vm.$data.results;
    expect(resultsRef).toEqual(validResults);
  });
});
