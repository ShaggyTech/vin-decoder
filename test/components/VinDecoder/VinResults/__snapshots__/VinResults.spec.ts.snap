// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`VinResults Component Tests component is hidden if rawResults prop is empty 1`] = `
<v-slide-y-transition-stub
  hideonleave="true"
  origin="top center 0"
>
  <!---->
</v-slide-y-transition-stub>
`;

exports[`VinResults Component Tests given rawResults, component renders the correct title and subtitle 1`] = `
<v-slide-y-transition-stub
  hideonleave="true"
  origin="top center 0"
>
  <v-card-stub
    class="results-card"
    loaderheight="4"
    maxwidth="600"
    raised="true"
    tag="div"
  >
    <v-card-title-stub
      class="mx-auto px-2 title"
    >
      <v-divider-stub />
       
      <span
        class="results-card__title mx-auto px-4"
      >
        
        2020 Volkswagen Jetta
        
      
      </span>
       
      <v-divider-stub />
    </v-card-title-stub>
     
    <v-card-subtitle-stub
      class="results-card__subtitle mx-auto px-2 title text-center"
    >
      
      TESTVIN
      
      <v-divider-stub />
    </v-card-subtitle-stub>
     
    <v-card-text-stub>
      <v-simple-table-stub
        class="results-card__table"
        fixedheader="true"
        height="300"
        max-width="600"
      />
    </v-card-text-stub>
  </v-card-stub>
</v-slide-y-transition-stub>
`;

exports[`VinResults Component Tests if loading prop is true, a skeleton loader is shown in place of the results card 1`] = `
<v-slide-y-transition-stub
  hideonleave="true"
  origin="top center 0"
>
  <v-skeleton-loader-stub
    class="results-loader"
    maxwidth="600"
    type="card-heading, text@7"
    types="[object Object]"
  />
</v-slide-y-transition-stub>
`;
